CorsixTH/Lua/humanoid_actions/seek_room.lua

```diff
@@ -246,115 +253,162 @@ local function action_seek_room_start(action, humanoid)
     return
   end
   -- Tries to find the room, if it is a diagnosis room, try to find any room in the diagnosis list.
-  local room = action_seek_room_find_room(action, humanoid)
+  if not humanoid.diagnosed or action.room_type == "research" then
+    local room = action_seek_room_find_room(action, humanoid)
+    -- if we have the room but not the staff we shouldn't seek out the room either
+    if room then
+      if humanoid.message then
+        TheApp.ui.bottom_panel:removeMessage(humanoid)
+      end
+      action_seek_room_goto_room(room, humanoid, action.diagnosis_room)
+      return
+    end
+  end
 
-  if room then
-    action_seek_room_goto_room(room, humanoid, action.diagnosis_room)
-  else
-    if not action.done_init then
-      action.done_init = true
-      action.must_happen = true
+  -- check we can treat the patient - and just shortcut the processing
+  local req = humanoid.hospital:checkDiseaseRequirements(humanoid.disease.id)
+  if humanoid.diagnosed and not req then
+    local room = action_seek_room_find_room(action, humanoid)
+    -- if we have the room but not the staff we shouldn't seek out the room either
+    if room then
+      if humanoid.message then
+        TheApp.ui.bottom_panel:removeMessage(humanoid)
+      end
+      action_seek_room_goto_room(room, humanoid, action.diagnosis_room)
+      return
+    end
+  end
+  -- we can't yet treat the patient, register callbacks
+  -- create message etc
+  if not action.done_init then
+    action.done_init = true
+    action.must_happen = true
 
-      local remove_callback = --[[persistable:action_seek_room_remove_callback]] function(rm)
-        if rm.room_info.id == "research" then
-          humanoid:updateMessage("research")
-        end
-      end -- End of remove_callback function
-      action.remove_callback = remove_callback
-      humanoid:registerRoomRemoveCallback(remove_callback)
+    local remove_callback = --[[persistable:action_seek_room_remove_callback]] function(rm)
+      humanoid:updateMessage("research")
+    end -- End of remove_callback function
+    action.remove_callback = remove_callback
+    humanoid:registerRoomRemoveCallback(remove_callback)
+
```

Comment:
If the room exists and it crashes or edited the room or deleted the room - we also want to update the message so removed the research check.